controllers/gateway/controller_test.go

```diff
@@ -37,6 +40,38 @@ func newController() *Controller {
 			Namespace:  common.DefaultControllerNamespace,
 			InstanceID: "argo-events",
 		},
+		TemplatesConfig: map[apicommon.EventSourceType]Template{
+			apicommon.WebhookEvent: Template{
+				Deployment: &corev1.PodTemplateSpec{
+					Spec: corev1.PodSpec{
+						Containers: []corev1.Container{
```

Comment:
Looking at the pod template of a gateway, I am realizing that we only need to configure the second container which is the actual events container. The first container which is `gateway-client` remains constant for all gateways and its behavior is fixed and there is no need to configure it separately. The only configuration for the `gateway-client` container can be to add `resources` to limit CPU and memory and we can make that default as part of the logic to construct the gateway pod template. 

So, the only thing we should let users configure is events container + pod related fields

If we store implements for different gateway types in controller configmap, then the user will lose the ability to configure different gateways of the same type in different ways.

Here is gateway CR spec I propose,
```

apiVersion: argoproj.io/v1alpha1
kind: Gateway
metadata:
  name: webhook-gateway
  labels:
    # the labels defined for the gateway CR will be applied on gateway deployment/pod and gateway service.
spec:
  type: webhook
  eventSourceRef:
    name: webhook-event-source
  # Optional
  template:
    # events container. we don't need configuration for gateway-client container.
    container:
      # container spec from core v1 K8s group.
      # user can specify container specific field as per their usecase.
      # we will have default values in controller logic.
    volumes:
      # list of volumes
    dnsPolicy:
      # Pod's DNS policy
    nodeSelector:
      # node selector to schedule the pod
    hostNetwork:
      # host network config
    securityContext:
      # security context
    affinity:
      # node affinity
    tolerations:
      # pod toleration
    priorityClassName:
    priority:
    dnsConfig:
  # Optional. should be used when user wants to expose the gateway pod.
  # the controller will create a ClusterIP service with `port` and `targetPort` as the values
  # in the list.
  # Whenever user adds or removes a port from this list, controller should update the ClusterIP service accordingly.
  service:
    ports:
      - 12000
      - 13000
      - 14000
  subscribers:
    http:
      - "http://webhook-sensor.argo-events.svc:9300/"


```

It is similar to how Argo Workflow spec is laid out.

Reply:
I agree not to let the user to configure `gateway-client` container, we can put its image in gateway-controller deployment spec as an ENV or argument. 

Where do you propose to put default event container config? Also, are you suggesting not to let the user to specify the image for event container? 

For service, I tend to not create it at all, because user always need to make change to it, or add new objects like `ingress` to it for public access. 

For gateway deployments like webhook, I suggest to let it listen to a fixed port (8080), I guess no user care about what port the container is listening to. For SSL support in gateway deployment, what we can do is put a self-signed CERT in the image on demand - this is what `argo-server` in workflow does right now, this is good enough to secure the communication from a load balancer to the POD. 

@VaibhavPage

Reply:
Also, do you see any requirements to expose properties like `dnsPolicy`, `nodeSelecor`... to the user at this time? If not, I would suggest not to expose them. I would rather not to expose them until we see lots of demands. Thoughts?

Reply:
Agreed. We can add those fields later as needed.